[Dynamic RPC] Allow for optional world_size argument in init_rpc

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74561 [Dynamic RPC] Add graceful shutdown for dynamic RPC members
* #74035 [Dynamic RPC] Allow existing ranks to communicate with newly joined ranks
* #73373 [Dynamic RPC] Allow newly joined ranks to communicate with existing ranks
* **#73372 [Dynamic RPC] Allow for optional world_size argument in init_rpc**

This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```

Tp run the added tests:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_dynamic_and_static_rpc_group`

Differential Revision: [D34621651](https://our.internmc.facebook.com/intern/diff/D34621651)

H-Huang committed:
[Dynamic RPC] Allow for optional world_size argument in init_rpc

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -376,8 +376,15 @@ TensorPipeAgent::TensorPipeAgent(
+    isStaticGroup_ = true;
+  } else {
+    // TODO: get current world size from store
+    isStaticGroup_ = false;
```
I assume all processes must all set `isStaticGroup_` to the same value? If so, let's also write this to store as well. If any process detected it's local `isStaticGroup_` is different from the one in store, it should error out.

H-Huang replied:
Created a new method `checkAndSetStaticGroup()` to handle this logic during initialization

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -51,19 +51,19 @@ def register_rendezvous_handler(scheme, handler):
 
 
-def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
+def rendezvous(url: str, rank: int = -1, world_size: Optional[int] = None, **kwargs):
```
curious, why do we need to convert it to optional? Will it work if we just use the default value -1 for dynamic world size?

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -164,7 +164,7 @@ def init_rpc(
         # keep it alive.
         global rendezvous_iterator
         rendezvous_iterator = dist.rendezvous(
```
We might need some renaming here, as `rendezvous()` is no longer a real rendezvous, it just creates a store. And if all we need is a store, we probably should just do that explicitly instead of going through the full `rendezvous` code path.

H-Huang replied:
Makes sense, this is not performing rendezvous anymore since nodes are not synchronizing themselves when constructing the store. There is some useful logic in rendezvous that takes `init_method` and validates arguments (`'rank`,`world_size`,`hostname`,`port`) based on the initialization type (`tcp`, `file`, `env`) which I was using before. I created a new method only to perform this validation and create the store `_create_store_for_rpc()` but I need to think about a good way to consolidate the repeated logic.

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -164,7 +164,7 @@ def init_rpc(
         global rendezvous_iterator
         rendezvous_iterator = dist.rendezvous(
-            rpc_backend_options.init_method, rank=rank, world_size=world_size
+            rpc_backend_options.init_method, rank=rank, world_size=world_size, init_rpc=True
```
curious, is `world_size=-1` sufficient to identify that this is a dynamic group? I would assume `rendezvous` / `store` does not need to now whether this is RPC or not?

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -243,6 +243,80 @@ def _tensorpipe_exchange_and_check_all_device_maps(
+            )
+        local_devices.update(key_set)
+
+    if not all(
```
is it possible to consolidate this with `_tensorpipe_validate_devices `?

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -243,6 +243,80 @@ def _tensorpipe_exchange_and_check_all_device_maps(
+                        f"device_maps = {device_maps}"
+                    )
+
+    gathered = api._all_gather([torch.cuda.device_count(), device_maps])
```
hmm, how does this work? It uses `_ALL_WORKER_NAMES` to do allgather?

H-Huang replied:
Removed this function in the next commit

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -260,55 +334,118 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+        token = f"TokenOnWorker{rank}"
+
+        while True:
+            returned = store.compare_set("init_rpc_token", "", token).decode()
```
nit: `init_rpc_token` this might be used for shutdown as well later, so it's more like an `authority`. Please feel free to come up with better names. :)

H-Huang replied:
Changed it to `RpcGroupManagementToken`, do you think that sounds okay? Or I can go with `authority` as well

mrshenli replied:
Yep `RpcGroupManagementToken` sounds good to me

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -260,55 +334,118 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+        while True:
+            returned = store.compare_set("init_rpc_token", "", token).decode()
+            if returned == token:
+                print(f"{rank} got token")
+                if torch.cuda.is_available():
+                    # It's necessary to initialize PyTorch CUDA states here (e.g.,
+                    # CUDACachingAllocator). If this is missing, we could hit errors like
+                    # "allocator not initialized", because other processes might send
+                    # CUDA-related RPC request to this process before user code in this
+                    # process initializes its PyTorch CUDA states.
+                    torch.cuda.init()
+                    device_count = torch.cuda.device_count()
+                else:
+                    device_count = 0
+
+                # Validate devices and device_maps locally for current rank
+                _tensorpipe_check_local_device_maps(name, rpc_backend_options)
```
we can do all of these before grabbing the token to minimize the contention for the critical section?

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```


[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```


[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```


[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34621651).

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -1,11 +1,14 @@
 
 import torch
 import torch.distributed as dist
+from ..rendezvous import _create_c10d_store  # noqa: F401
```
does it make sense to import `_create_c10d_store` to `distributed/__init__.py` so that we can directly call `dist._create_c10d_store` in this file? See example below

https://github.com/pytorch/pytorch/blob/a39e8e8f5e232cbc74f94fe2440be100823602ce/torch/distributed/__init__.py#L61-L67

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -195,23 +201,48 @@ def _validate_rpc_args(backend, store, name, rank, world_size, rpc_backend_optio
+                     in filter(None, result.query.split("&")))  # type: ignore[misc, arg-type, var-annotated]
+        if scheme == "file":
+            ValueError("File initialization is not supported for Dynamic RPC")
+        else:
+            hostname = _get_env_or_raise("MASTER_ADDR")
+            port = int(_get_env_or_raise("MASTER_PORT"))
+        if rank == -1:
+            if "rank" in query:
+                rank = int(query["rank"])
+            else:
+                rank = int(_get_env_or_raise("RANK"))
+        return _create_c10d_store(hostname, port, rank, None, backend_options.rpc_timeout)
```
Can we consolidate duplicated code with `rendezvous()`? E.g., we can have a helper function that wraps all the common code and the both this function and `rendezvous()` call into that function.

H-Huang replied:
Thanks! I merged this method to also use the rendezvous handlers

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -162,11 +165,14 @@ def init_rpc(
+            )
+            store, _, _ = next(rendezvous_iterator)
+        else:
+            store = _create_store_for_rpc(rpc_backend_options, rank)
```
This looks a bit inconsistent with the `if` clause. Does it make sense to call it `dist.join()` or `dist.connect()`?

And looks like `_create_store_for_rpc` does not seem it contains RPC specific code. If that's the case, we can keep it in `rendezvous.py` for now and also pass `rpc_backend_options.init_method` and timeout explicitly to it.

H-Huang replied:
moved it into rendezvous and creates the store based on backend_options

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -277,38 +295,88 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+                )
 
+                try:
+                    # TODO: Notify all workers in group this rank has joined and set devices and reverse_device_map
```
Looks like without this, we cannot reasonable coverage on tests. Let's wait till this is addressed in the next PR before landing.

H-Huang replied:
I added tests in `rpc_test` to cover initialization. There will be tests in remaining PRs which enables communication and can be tested.

mrshenli reviewed (commented):
I recall in one of our previous discussion, you mentioned that if B joined after A, and B discovered B. Then if B sends an RPC to A, and actually can send the response to successfully, even if A wasn't notified about B's existence yet. And we were thinking it is because TensorPipe used the same pipe for request and response. @lw is this correct assumption?

If above is expected, shall we add a test for the above scenario?

mrshenli reviewed (commented):
Lint failure is real

lw commented:
> And we were thinking it is because TensorPipe used the same pipe for request and response. @lw is this correct assumption?

I don't know what's the context for this, and this doesn't really depend on TensorPipe but on how the RPC agent chooses to use TensorPipe. Last time I checked this was the case:
- The A->B pipe was used for A sending requests to B and B sending responses to A
- The B->A pipe (which is separate!) was used for B sending requests to A and A sending responses to B

H-Huang commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -354,11 +354,38 @@ void TensorPipeAgent::prepareNames() {
     const c10::intrusive_ptr<::c10d::Store>& store,
     std::string selfName,
     worker_id_t selfId,
-    int worldSize,
```
There is no reasoning other than coding style. Since a magic number like -1 or 0 could be used to designate dynamic membership it would be confusing to users; a null value is a bit more explicit. We could also handle and values <= 0 as dynamic.

H-Huang commented on torch/distributed/rendezvous.py:
```diff
@@ -151,6 +151,10 @@ def _create_c10d_store(hostname, port, rank, world_size, timeout) -> Store:
     if not 0 <= port < 2**16:
         raise ValueError(f"port must have value from 0 to 65535 but was {port}.")
 
+    # if world_size is None, then there is a non-fixed number of stores
```
Yep, thanks!

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```

Tp run the added tests:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D34621651](https://our.internmc.facebook.com/intern/diff/D34621651)

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -354,11 +354,38 @@ void TensorPipeAgent::prepareNames() {
+            isStaticGroup_,
+            storeIsStaticGroup))
+  } else {
+    std::string isStaticGroupStr;
+    isStaticGroupStr = isStaticGroup_ ? "true" : "false";
+    std::vector<uint8_t> isStaticGroupVec(
+        (uint8_t*)isStaticGroupStr.c_str(),
+        (uint8_t*)isStaticGroupStr.c_str() + isStaticGroupStr.length());
+    store->set(isStaticGroupKey, isStaticGroupVec);
```
Correct compareSet does not have a good way of verifying whether the value was set or previously existed. Using the regular set() makes the logic simpler.

H-Huang replied:
Thanks, there is a race condition because this is used in static TP initialization. Curious, what did you envision to be the purpose of this method? I assumed this was to do a sanity check to make sure that there are not both static and dynamic RPC groups being initialized. If so, I think I can use compareSet() to fix the race condition.

mrshenli replied:
>  Curious, what did you envision to be the purpose of this method? I assumed this was to do a sanity check to make sure that there are not both static and dynamic RPC groups being initialized.

Yep, exactly

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -376,8 +403,17 @@ TensorPipeAgent::TensorPipeAgent(
+      shutdownStore_("shutdown", store) {
+  if (worldSize.has_value()) {
+    worldSize_ = worldSize.value();
+    isStaticGroup_ = true;
```
nit: `isStaticGroup_` can be `const bool`?

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -50,6 +50,13 @@ def register_rendezvous_handler(scheme, handler):
+# converted into {"rank": 0, "world_size": 1}
+def _query_to_dict(query):
+    query_dict: Dict[str, Union[int, str]] = dict(
+        cast(Tuple[str, Union[int, str]], pair.split("=")) for pair in cast(Iterable[str], filter(None, query.split("&")))
```
curious, any reason that we need a filter given the condition is `None`?

mrshenli replied:
Can this be simplified to the following?

```python
dict(cast(Tuple[str, Union[int, str]], pair.split("=")) for pair in query.split("&"))
```

mrshenli replied:
Regarding `Tuple[str, Union[int, str]]`, when will the key be int? For input `"rank=0&world_size=1" `, looks like the output is the following?

```python
{'rank': '0', 'world_size': '1'}
```

H-Huang replied:
> curious, any reason that we need a filter given the condition is None?

The None filters out any empty values, such as if the query contained `"rank=0&world_size=1&"` (note the extra &)

> Can this be simplified to the following? 
```python 
dict(cast(Tuple[str, Union[int, str]], pair.split("=")) for pair in query.split("&"))
```

Need the cast to Iterable[str] or else mypy complains `torch/distributed/rendezvous.py:56:38: error: Need type annotation for 'pair'  [var-annotated]`

> Regarding Tuple[str, Union[int, str]], when will the key be int? 

Good point, it will not be int. `Tuple[str, str]` is sufficient.

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -88,6 +93,33 @@ def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
+        # if the world_size env variable is not present then it is a dynamic group
+        world_size = int(os.environ.get("WORLD_SIZE", world_size))
+
+    query_dict = _query_to_dict(result.query)
+    if rank != -1:
+        query_dict["rank"] = rank
```
Could you please elaborate the logic here? Would I be correct that we currently require applications to specify a rank regardless of whether this is dynamic grouping or not?

mrshenli replied:
BTW, would I be correct that we need a TODO item to automatically assign ranks? Because with dynamic world size, a developer might not know the rank for the process?

H-Huang replied:
> Could you please elaborate the logic here? Would I be correct that we currently require applications to specify a rank regardless of whether this is dynamic grouping or not?

Yes rank is always required. However, it does not necessarily need to be passed into `init_rpc` as `RANK` and `WORLD_SIZE` can be set in environment variables and used during `env://` initialization. Therefore we run into the issue of needing to check that environment variables for the rank value and setting it to that. If rank is not set in either `init_rpc` or `env` then it will be missing from the reformatted query and will throw an error in `_rendezvous_handlers` with error `rank parameter is missing`

H-Huang replied:
> BTW, would I be correct that we need a TODO item to automatically assign ranks? Because with dynamic world size, a developer might not know the rank for the process?

Yes, an additional feature would be to assign ranks internally so users do not need to assign the ranks themselves.

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -88,6 +93,33 @@ def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
+    if result.scheme == "env":
+        rank = os.environ.get("RANK", rank)
+        # world_size default set to -1
+        # if the world_size env variable is not present then it is a dynamic group
```
Could you please write a complete explanation of what condition will trigger dynamic group?

H-Huang replied:
Added additional explanation in `__init__.py` The complete explanation is just that if `world_size` is not set in `init_rpc` and also not set in environment variable `WORLD_SIZE`, then that triggers dynamic group initialization

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -88,6 +93,33 @@ def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
+        query_dict["rank"] = rank
+    query_dict["world_size"] = world_size
+
+    result = result._replace(
+        query="{}".format(
+            "&".join(["{}={}".format(k, v) for k, v in query_dict.items()])
+        )
+    )
+
+    url = urlunparse(result)
+    if result.scheme not in _rendezvous_handlers:
+        raise RuntimeError("No handler for {}://".format(result.scheme))
```
there are still some code duplication with the `rendezvous` method. Let's create an issue to track this.

mrshenli replied:
add #74440

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -88,6 +93,33 @@ def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
+    )
+
+    url = urlunparse(result)
+    if result.scheme not in _rendezvous_handlers:
```
does FileStore support dynamic group or just TCPStore support that?

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -151,6 +181,10 @@ def _create_c10d_store(hostname, port, rank, world_size, timeout) -> Store:
 
+    # if world_size is None then there is a non-fixed number of members
+    if not world_size:
+        world_size = -1
```
(can be done in followup PR): If it is designated to use `optional<int>` and `None` to indicate dynamic group (as in TensorPipeAgent), let's make it consistent everywhere (i.e., modify TCPStore accordingly).

mrshenli replied:
Hold on, looks like TCPStore already accepts None?

https://github.com/pytorch/pytorch/blob/f0db315efeef27a2d8760cec4573bcbe40854b01/torch/csrc/distributed/c10d/TCPStore.cpp#L964

H-Huang replied:
It does but the pybind definition does not. I added an issue to track. https://github.com/pytorch/pytorch/issues/74488

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4320,6 +4320,47 @@ def test_init_rpc_twice(self):
+        )
+
+        # TODO: shutdown is ungraceful, not yet fully implemented
+        rpc.shutdown(graceful=False)
```
Will this introduce any flakiness?

H-Huang replied:
Initially I thought it wouldn't since there is no communication and this test is only testing initialization. But now I can think of a case where Rank 0 shutdowns before another rank initializations, e.g. Rank 1. This would cause Rank 1 to hang and eventually error out.

H-Huang replied:
I added a sync before shutdown to solve this. Will remove once graceful shutdown is implemented.

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```

Tp run the added tests:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D34621651](https://our.internmc.facebook.com/intern/diff/D34621651)

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```

Tp run the added tests:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D34621651](https://our.internmc.facebook.com/intern/diff/D34621651)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34621651).

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -354,11 +354,41 @@ void TensorPipeAgent::prepareNames() {
+  TORCH_CHECK(
+      returnedVal == isStaticGroupStr,
+      fmt::format(
+          "RPC group behavior is different than expected. isStaticGroup_ is initialized with {} while in store it is {}",
```
Is it possible to add a test to cover this? I assume not, because the static code path would actually create a process group before creating TP agent instances?

mrshenli reviewed (approved):
LGTM! Thanks a lot for making steady progress! There are several irrelevant fx test failures in CI. Please rebase to avoid that.

H-Huang committed:
Update on "[Dynamic RPC] Allow for optional world_size argument in init_rpc"


This PR which allows for optional `world_size` argument in init_rpc. This makes changes in rendezvous to allow for `NoneType` for world_size and creates a new code path when initializing TensorPipe agent for init_rpc. The TensorPipe agent is protected by a critical section enforced using the store, so that only one node can create a TPAgent at a time.
This PR does not yet enable RPC commands between ranks.

Previously:
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", world_size=1, rank=0)
```

Now (only rank is needed):
```python
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '29500'
init_rpc("worker0", rank=0)
```

Tp run the added tests:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D34621651](https://our.internmc.facebook.com/intern/diff/D34621651)

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -354,11 +354,41 @@ void TensorPipeAgent::prepareNames() {
+  } else {
+    returnedVec = store->compareSet(
+        isStaticGroupKey, std::vector<uint8_t>(), isStaticGroupVec);
+  }
```
I was able to remove the check call and only use compareSet, so this now works with concurrency.

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34621651).

H-Huang commented:
@pytorchbot retest this please
